operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java

```diff
@@ -84,26 +86,35 @@
         // TODO use caches in secondary resources; this is a workaround for https://github.com/java-operator-sdk/java-operator-sdk/issues/830
         // KeycloakDeployment deployment = new KeycloakDeployment(client, config, kc, context.getSecondaryResource(Deployment.class).orElse(null));
         var kcDeployment = new KeycloakDeployment(client, config, kc, null);
-        kcDeployment.updateStatus(statusBuilder);
+        var deploymentReady = kcDeployment.updateStatus(statusBuilder);
         kcDeployment.createOrUpdateReconciled();
 
         var kcService = new KeycloakService(client, kc);
-        kcService.updateStatus(statusBuilder);
+        var serviceReady = kcService.updateStatus(statusBuilder);
         kcService.createOrUpdateReconciled();
         var kcDiscoveryService = new KeycloakDiscoveryService(client, kc);
-        kcDiscoveryService.updateStatus(statusBuilder);
+        var discoveryServiceReady = kcDiscoveryService.updateStatus(statusBuilder);
         kcDiscoveryService.createOrUpdateReconciled();
 
         var status = statusBuilder.build();
 
         Log.info("--- Reconciliation finished successfully");
+        var ready = deploymentReady && serviceReady && discoveryServiceReady;
 
         if (status.equals(kc.getStatus())) {
-            return UpdateControl.noUpdate();
+            if (ready) {
+                return UpdateControl.noUpdate();
+            } else {
+                return UpdateControl.<Keycloak>noUpdate().rescheduleAfter(10, TimeUnit.SECONDS);
+            }
         }
         else {
             kc.setStatus(status);
-            return UpdateControl.updateStatus(kc);
+            if (ready) {
+                return UpdateControl.updateStatus(kc);
+            } else {
+                return UpdateControl.updateStatus(kc).rescheduleAfter(10, TimeUnit.SECONDS);
+            }
```

Comment:
I presume we need these ugly `if`s because of `generationAwareEventProcessing = false`. Correct?
So we're effectively managing the reconciliation loop and relying on time checks instead of events? That means the CRs are not watched any more?

If all the answers are "yes", I'd rather see `rescheduleAfter` even when `ready == true`. What if the deployment fails suddenly? The operator would not know.

Reply:
> I presume we need these ugly ifs because of `generationAwareEventProcessing = false`. Correct?

No, this is just hardening the logic for retrying the reconciliation.

> So we're effectively managing the reconciliation loop and relying on time checks instead of events? That means the CRs are not watched any more?

No, this change simply means: "if we know that there is additional processing to be done from the operator we install a fail-safe temporal fallback".
You can read more about [here](https://javaoperatorsdk.io/docs/features) in the section "Retry and Rescheduling and Event Handling Common Behavior"

> What if the deployment fails suddenly? The operator would not know.

We should keep relying on events, and we can possibly remove those safe-nets later on in the development.

Reply:
Thanks for the clarification. In that case, could we do something like this?

```java
UpdateControl<Keycloak> ret;

        if (status.equals(kc.getStatus())) {
            ret = UpdateControl.noUpdate();
        }
        else {
            ret = UpdateControl.updateStatus(kc);
        }
        
        if (status.getConditions().stream()
                .anyMatch(c -> c.getType().equals(KeycloakStatusCondition.READY) && !c.getStatus())) {
            ret.rescheduleAfter(10, TimeUnit.SECONDS);
        }

        return ret;
```

I hate that we are now introducing additional logic for detecting ready state while we already have this information.

Reply:
Unfortunately the proposed change invalidates the fail over logic:
```java
status.getConditions().stream()
        .anyMatch(c -> c.getType().equals(KeycloakStatusCondition.READY)
```
Won't detect the fact that a [`service` needs to be re-created](https://github.com/keycloak/keycloak/pull/10653/files#diff-6a6cb293bff38002e801b4adaf350b0bb8ead4599ad52903b654fde6ddcb4350R88).

I anyhow refactored the logic to be a little nicer to the eyes.

@vmuzikar those fail over logics and the workarounds needs to be audited and reviewed as soon as we have a proper fix from the JOSDK.

Reply:
this should be resolved now @vmuzikar